app/scripts/modules/core/src/presentation/hooks/useDataSource.hook.ts

```diff
@@ -27,6 +33,10 @@ export const useDataSource = <T>(dataSource: ApplicationDataSource<T>): IDataSou
 
   return {
     data,
+    status,
```

Comment:
AFAICT `status` will end up being a nested `IFetchStatus` object that has a duplicate copy of `data` which may be out of sync with the one sourced from `data$` depending on the two streams' activity. Can you lift anything you specifically want out of `IFetchStatus` into a top level field? I'd prefer to see this result object remain pretty flat anyway.

Reply:
+1 I think we could just destructure `status` here, i.e.: 
`return { ...status, refresh }`

In general I prefer to avoid flag fatigue by using enums (when flags are mutually exclusive, i.e., `loaded, loading, loadFailure`).

Reply:
Didn't realize status was a nested object. Also +1 on just destructuring it.

> `loaded`, `loading`, `loadFailure`

Unfortunately they are not all mutually exclusive because `loaded` tracks if it as ever loaded:

https://github.com/spinnaker/deck/blob/b83fcf77fe6d876afd49756d4f6b38922688554a/app/scripts/modules/core/src/application/service/applicationDataSource.ts#L327-L329

I didn't like the flag fatigue either, but I also didn't want this logic ☝️ to just be duplicated everywhere. I'm happy to drop the latter flags, but I think we'll need to keep `loaded`.

Reply:
Gotcha.  
Maybe include `loaded` in `IFetchStatus`?  Maybe give it a less ambiguous name like `hasEverLoaded` or `ready` or something?

Reply:
very 👍 to adding a totally new property to `IFetchStatus` _if we need to_ that's more clear. 

Only question I have before we go down that route: would checking for `lastRefresh > 0` resolve your concern about duplication of logic? Based on reading the code I _think_ `lastRefresh` remains set to `0` until the first fetch has successfully completed. That's similar to how we do 'has ever logic' with `ILatestPromiseResult` which has a monotonically incrementing `requestId`. If that's still going to result in too much spaghetti in this case then let's do a standalone flag.

Reply:
I'd prefer not to invent a new way of making that calculation. It's even less obvious that checking `lastRefresh > 0` actually means "has ever loaded" and they are equivalent today but we can't guarantee that will hold true in the future (for example, if one decided to update lastRefresh even for errors 🤷‍♂ ).

I'd also rather keep the name `loaded` to be consistent with what it is and means on `dataSource.loaded`. I just don't want to introduce yet another new thing that actually means the same thing and cause someone to think

- what is the difference between `loaded` and `hasEverLoaded`??
- or worse: does `loaded` not actually mean "has ever loaded"? what does it mean then?

It doesn't sound like there are any strong objections, but please let me know if there are. Commit with updates coming shortly. And 🙏 thank you for the thoughtful reviews!

Reply:
that seems extremely reasonable, I am happy to avoid folks getting confused trying to parse the data source API given it's already a little hard to wrap one's head around